services/opentsdb/handler.go

```diff
@@ -0,0 +1,180 @@
+package opentsdb
+
+import (
+	"bufio"
+	"compress/gzip"
+	"encoding/json"
+	"errors"
+	"io"
+	"log"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/influxdb/influxdb"
+	"github.com/influxdb/influxdb/cluster"
+	"github.com/influxdb/influxdb/tsdb"
+)
+
+type Handler struct {
+	Database        string
+	RetentionPolicy string
+
+	PointsWriter interface {
+		WritePoints(p *cluster.WritePointsRequest) error
+	}
+
+	Logger *log.Logger
+}
+
+func (h *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	switch r.URL.Path {
+	case "/api/metadata/put":
+		w.WriteHeader(http.StatusNoContent)
+	case "/api/put":
+		h.servePut(w, r)
+	default:
+		http.NotFound(w, r)
+	}
+}
+
+// ServeHTTP implements OpenTSDB's HTTP /api/put endpoint
+func (h *Handler) servePut(w http.ResponseWriter, r *http.Request) {
+	defer r.Body.Close()
+
+	// Require POST method.
+	if r.Method != "POST" {
+		http.Error(w, http.StatusText(http.StatusMethodNotAllowed), http.StatusMethodNotAllowed)
+		return
+	}
+
+	// Wrap reader if it's gzip encoded.
+	var br *bufio.Reader
+	if r.Header.Get("Content-Encoding") == "gzip" {
+		zr, err := gzip.NewReader(r.Body)
+		if err != nil {
+			http.Error(w, "could not read gzip, "+err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		br = bufio.NewReader(zr)
+	} else {
+		br = bufio.NewReader(r.Body)
+	}
+
+	// Lookahead at the first byte.
+	f, err := br.Peek(1)
+	if err != nil || len(f) != 1 {
+		http.Error(w, "peek error: "+err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	// Peek to see if this is a JSON array.
+	var multi bool
+	switch f[0] {
+	case '{':
+	case '[':
+		multi = true
+	default:
+		http.Error(w, "expected JSON array or hash", http.StatusBadRequest)
+		return
+	}
+
+	// Decode JSON data into slice of points.
+	dps := make([]point, 1)
+	if dec := json.NewDecoder(br); multi {
+		if err = dec.Decode(&dps); err != nil {
+			http.Error(w, "json array decode error", http.StatusBadRequest)
+			return
+		}
+	} else {
+		if err = dec.Decode(&dps[0]); err != nil {
+			http.Error(w, "json object decode error", http.StatusBadRequest)
+			return
+		}
+	}
+
+	// Convert points into TSDB points.
+	points := make([]tsdb.Point, 0, len(dps))
+	for i := range dps {
+		p := dps[i]
+
+		// Convert timestamp to Go time.
+		// If time value is over a billion then it's microseconds.
+		var ts time.Time
+		if p.Time < 10000000000 {
+			ts = time.Unix(p.Time, 0)
+		} else {
+			ts = time.Unix(p.Time/1000, (p.Time%1000)*1000)
+		}
+
+		points = append(points, tsdb.NewPoint(p.Metric, p.Tags, map[string]interface{}{"value": p.Value}, ts))
+	}
+
+	// Write points.
+	if err := h.PointsWriter.WritePoints(&cluster.WritePointsRequest{
+		Database:         h.Database,
+		RetentionPolicy:  h.RetentionPolicy,
+		ConsistencyLevel: cluster.ConsistencyLevelOne,
```

Comment:
I just want to convert the old server over to a service right now. We can add consistency level separately.